sycl/include/sycl/types.hpp

```diff
@@ -2085,298 +2085,119 @@ __SYCL_RELLOGOP(||)
-  };
-
-#define __SYCL_DECLARE_BOOL_CONVERTER(num)                                     \
-  template <> class BaseCLTypeConverter<bool, num> {                           \
-  public:                                                                      \
-    using DataType = detail::select_apply_cl_t<                                \
-        bool, __SYCL_GET_CL_TYPE(char, num), __SYCL_GET_CL_TYPE(short, num),   \
-        __SYCL_GET_CL_TYPE(int, num), __SYCL_GET_CL_TYPE(long, num)>;          \
-  };
-
-#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
-#define __SYCL_DECLARE_BYTE_CONVERTER(num)                                     \
-  template <> class BaseCLTypeConverter<std::byte, num> {                      \
-  public:                                                                      \
-    using DataType = __SYCL_GET_CL_TYPE(uchar, num);                           \
-  };
-#endif
-#define __SYCL_DECLARE_HALF_CONVERTER(base, num)                               \
-  template <> class BaseCLTypeConverter<base, num> {                           \
-  public:                                                                      \
-    using DataType = __SYCL_GET_CL_HALF_TYPE(base, num);                       \
-  };
 
-#define __SYCL_DECLARE_SCALAR_SCHAR_CONVERTER                                  \
-  template <> class BaseCLTypeConverter<schar, 1> {                            \
-  public:                                                                      \
-    using DataType = schar;                                                    \
-  };
-
-#define __SYCL_DECLARE_SCALAR_BOOL_CONVERTER                                   \
-  template <> class BaseCLTypeConverter<bool, 1> {                             \
-  public:                                                                      \
-    using DataType = bool;                                                     \
-  };
+// Vectors of size 1 are handled separately and therefore 1 is not included in
+// the check below.
+constexpr bool isValidVectorSize(int N) {
+  return (N >= 2 && N <= 4) || N == 8 || N == 16;
+}
+template <typename T, int N, typename V> struct VecStorage {
+  static_assert(
+      isValidVectorSize(N) || N == 1,
+      "Incorrect number of elements for sycl::vec: only 1, 2, 3, 4, 8 "
+      "or 16 are supported");
+  static_assert(!std::is_same_v<V, void>, "Incorrect data type for sycl::vec");
+};
 
-#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
-#define __SYCL_DECLARE_SCALAR_BYTE_CONVERTER                                   \
-  template <> class BaseCLTypeConverter<std::byte, 1> {                        \
-  public:                                                                      \
-    using DataType = __SYCL_GET_SCALAR_CL_TYPE(uchar);                         \
+#ifdef __SYCL_USE_EXT_VECTOR_TYPE__
+template <typename T, int N> struct VecStorageImpl {
+  using DataType = T __attribute__((ext_vector_type(N)));
```

Comment:
This shouldn't be an ABI break, because we use the same underlying types now. Essentially, we use (and still do so in this PR) `ext_vector_type` when it is available and fallback to types defined in `CL/OpenCL.h` otherwise.

The only change here is the mechanism of how exactly mapping is done.

However, ABI is indeed different depending on which host compiler is used, because of lack (or presence) of `ext_vector_type` extension. But again, this PR introduces no changes to that